File: Sources/BranchSDK/BranchConfiguration.m
//
//  BranchConfiguration.m
//  BranchSDK
//
//  Created by Brandon Boothe on 7/21/26.
//

#import "BranchConfiguration.h"
#import "BranchConfiguration+Private.h"
#import "NSError+Branch.h"

// Defaults mirror BNCPreferenceHelper (DEFAULT_TIMEOUT, DEFAULT_RETRY_COUNT, DEFAULT_RETRY_INTERVAL).
static const NSTimeInterval BranchConfigurationDefaultNetworkTimeout = 5.5;
static const NSInteger      BranchConfigurationDefaultRetryCount     = 3;
static const NSTimeInterval BranchConfigurationDefaultRetryInterval  = 0;

// Mirrors BNCPreferenceHelper's DEFAULT_THIRD_PARTY_APIS_TIMEOUT (500ms).
static const NSTimeInterval BranchConfigurationDefaultThirdPartyAPIsWaitTime = 0.5;

// Network timeout ceiling, matching the Android builder's 60s cap.
static const NSTimeInterval BranchConfigurationMaxNetworkTimeout = 60;

// Ceiling for the third-party API wait time, matching +[Branch setSDKWaitTimeForThirdPartyAPIs:].
static const NSTimeInterval BranchConfigurationMaxThirdPartyAPIsWaitTime = 10;

@interface BranchConfiguration ()
@property (nonatomic, copy, readwrite) NSString *branchKey;
@property (nonatomic, assign, readwrite) BOOL logLevelWasSet;
@property (nonatomic, assign, readwrite) BOOL euEndpointWasSet;
@property (nonatomic, strong) NSMutableArray<NSString *> *mutableAllowedSchemes;
@property (nonatomic, strong) NSMutableArray<NSString *> *mutableUrlPatternsToIgnore;
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSString *> *mutableRequestMetadata;
@end

@implementation BranchConfiguration

- (instancetype)initWithKey:(NSString *)branchKey {
    self = [super init];
    if (!self) return self;

    _branchKey = [branchKey copy];

    // Identity & environment
    _testMode = NO;
    _apiUrl = nil;
    _safeTrackAPIUrl = nil;
    _cdnBaseUrl = nil;
    _euEndpoint = NO;
    _euEndpointWasSet = NO;

    // Logging
    _logLevel = BranchLogLevelError;
    _logLevelWasSet = NO;
    _loggingCallback = nil;
    _requestTracingCallback = nil;

    // Network
    _networkTimeout = BranchConfigurationDefaultNetworkTimeout;
    _retryCount = BranchConfigurationDefaultRetryCount;
    _retryInterval = BranchConfigurationDefaultRetryInterval;
    _remoteInterface = nil;
    _thirdPartyAPIsWaitTime = BranchConfigurationDefaultThirdPartyAPIsWaitTime;

    // Privacy & attribution
    _attributionLevel = nil;
    _limitFacebookAttribution = NO;
    _adNetworkCalloutsDisabled = NO;
    _dmaParameters = nil;

    // Collections
    _mutableAllowedSchemes = [NSMutableArray array];
    _mutableUrlPatternsToIgnore = [NSMutableArray array];
    _mutableRequestMetadata = [NSMutableDictionary dictionary];

    // Open tracking
    _automaticOpenEvents = YES;

    // Pasteboard
    _checkPasteboardOnInstall = NO;

    // App Clip
    _appClipAppGroup = nil;

    // Debugging
    _deepLinkDebugParams = nil;

    return self;
}

#pragma mark - Logging

// BranchLogLevel is a non-optional enum whose first case, BranchLogLevelVerbose, is 0, so the value
// alone cannot distinguish "never set" from "set to Verbose". Record the assignment instead.
- (void)setLogLevel:(BranchLogLevel)logLevel {
    _logLevel = logLevel;
    _logLevelWasSet = YES;
}

#pragma mark - Identity & environment

// A BOOL cannot distinguish "never set" from "set to NO", so an unset euEndpoint would leave
// -[Branch useEUEndpoints] as the only way to reach EU routing and no way at all to turn it back
// off. Record the assignment so `euEndpoint = NO` is honored as an explicit choice.
- (void)setEuEndpoint:(BOOL)euEndpoint {
    _euEndpoint = euEndpoint;
    _euEndpointWasSet = YES;
}

#pragma mark - Factory helpers

+ (instancetype)debug:(NSString *)branchKey {
    BranchConfiguration *config = [[self alloc] initWithKey:branchKey];
    config.logLevel = BranchLogLevelVerbose;
    config.testMode = YES;
    return config;
}

+ (instancetype)production:(NSString *)branchKey {
    BranchConfiguration *config = [[self alloc] initWithKey:branchKey];
    config.logLevel = BranchLogLevelWarning;
    return config;
}

+ (instancetype)compliance:(NSString *)branchKey {
    BranchConfiguration *config = [[self alloc] initWithKey:branchKey];
    config.attributionLevel = BranchAttributionLevelMinimal;
    return config;
}

#pragma mark - Collection accessors

- (NSArray<NSString *> *)allowedSchemes {
    return [self.mutableAllowedSchemes copy];
}

- (NSArray<NSString *> *)urlPatternsToIgnore {
    return [self.mutableUrlPatternsToIgnore copy];
}

- (NSDictionary<NSString *, NSString *> *)requestMetadata {
    return [self.mutableRequestMetadata copy];
}

- (void)addAllowedScheme:(NSString *)scheme {
    if (scheme) {
        [self.mutableAllowedSchemes addObject:scheme];
    }
}

- (void)addMetadataWithKey:(NSString *)key value:(NSString *)value {
    if (key && value) {
        self.mutableRequestMetadata[key] = value;
    }
}

// The collection setters replace the backing store outright rather than merging, so assigning after
// -addAllowedScheme: / -addMetadataWithKey:value: discards those entries. Assigning nil clears the
// collection instead of leaving a nil store, keeping the readonly-style getters non-nil.
- (void)setAllowedSchemes:(nullable NSArray<NSString *> *)allowedSchemes {
    self.mutableAllowedSchemes = [allowedSchemes mutableCopy] ?: [NSMutableArray array];
}

- (void)setRequestMetadata:(nullable NSDictionary<NSString *, NSString *> *)requestMetadata {
    self.mutableRequestMetadata = [requestMetadata mutableCopy] ?: [NSMutableDictionary dictionary];
}

- (void)setUrlPatternsToIgnore:(NSArray<NSString *> *)urlPatternsToIgnore {
    self.mutableUrlPatternsToIgnore = [urlPatternsToIgnore mutableCopy] ?: [NSMutableArray array];
}

#pragma mark - Validation

- (BOOL)validate:(NSError *_Nullable *_Nullable)error {
    if (self.branchKey.length == 0) {
        return [self failValidation:@"Branch key cannot be empty. Get your key from dashboard.branch.io/settings."
                              error:error];
    }
    if (self.networkTimeout <= 0) {
        return [self failValidation:[NSString stringWithFormat:@"Network timeout must be a positive number of seconds (got %.2f).", self.networkTimeout]
                              error:error];
    }
    if (self.networkTimeout > BranchConfigurationMaxNetworkTimeout) {
        return [self failValidation:[NSString stringWithFormat:@"Network timeout cannot exceed 60 seconds (got %.2f).", self.networkTimeout]
                              error:error];
    }
    if (self.retryCount < 0) {
        return [self failValidation:[NSString stringWithFormat:@"Retry count must be >= 0 (got %ld).", (long)self.retryCount]
                              error:error];
    }
    if (self.retryInterval < 0) {
        return [self failValidation:[NSString stringWithFormat:@"Retry interval must be >= 0 seconds (got %.2f).", self.retryInterval]
                              error:error];
    }
    if (self.thirdPartyAPIsWaitTime <= 0 || self.thirdPartyAPIsWaitTime > BranchConfigurationMaxThirdPartyAPIsWaitTime) {
        return [self failValidation:[NSString stringWithFormat:@"Third-party APIs wait time must be > 0 and <= 10 seconds (got %.2f).", self.thirdPartyAPIsWaitTime]
                              error:error];
    }
    if (self.remoteInterface && ![self.remoteInterface conformsToProtocol:@protocol(BNCNetworkServiceProtocol)]) {
        return [self failValidation:[NSString stringWithFormat:@"remoteInterface class '%@' must conform to BNCNetworkServiceProtocol.",
                                     NSStringFromClass(self.remoteInterface)]
                              error:error];
    }
    if (self.apiUrl && !([self.apiUrl hasPrefix:@"http://"] || [self.apiUrl hasPrefix:@"https://"])) {
        return [self failValidation:[NSString stringWithFormat:@"A custom apiUrl must either have a prefix of http:// or https:// (got '%@').", self.apiUrl]
                              error:error];
    }
    if (self.safeTrackAPIUrl && !([self.safeTrackAPIUrl hasPrefix:@"http://"] || [self.safeTrackAPIUrl hasPrefix:@"https://"])) {
        return [self failValidation:[NSString stringWithFormat:@"A custom safeTrackAPIUrl must either have a prefix of http:// or https:// (got '%@').", self.safeTrackAPIUrl]
                              error:error];
    }
    return YES;
}

// Fills the caller's out-param, when supplied, with a BNCInvalidConfigurationError carrying `message`
// as its failure reason. Always returns NO so each check above can `return [self failValidation:…]`.
- (BOOL)failValidation:(NSString *)message error:(NSError *_Nullable *_Nullable)error {
    NSError *validationError = [NSError branchErrorWithCode:BNCInvalidConfigurationError localizedMessage:message];
    [[BranchLogger shared] logError:message error:validationError];
    if (error) {
        *error = validationError;
    }
    return NO;
}

@end

File: Sources/BranchSDK_Swift/BranchConfiguration.swift
//
//  BranchConfiguration.swift
//  BranchSDK
//
//  Swift port of the former Objective-C BranchConfiguration.
//
//  `@objc(BranchConfiguration)` and per-member @objc names are pinned so existing
//  Objective-C call sites (`+[Branch initialize:]`) and unit tests keep working unchanged.
//

import Foundation

#if SWIFT_PACKAGE
import BranchObjCSDK
#endif

/// `BranchConfiguration` collects every pre-init decision for the Branch SDK into a single object that
/// is handed to `+[Branch initialize:]`. It is the iOS counterpart of the Android
/// `BranchConfiguration.Builder`.
///
/// Create it with a Branch key, set the properties you care about, then pass it to `+[Branch initialize:]`.
/// Once `initialize:` has been called, mutating the configuration has no effect — `initialize:` reads a
/// snapshot of the values.
@objc(BranchConfiguration)
public class BranchConfiguration: NSObject {

    // Defaults mirror BNCPreferenceHelper (DEFAULT_TIMEOUT, DEFAULT_RETRY_COUNT, DEFAULT_RETRY_INTERVAL).
    private static let defaultNetworkTimeout: TimeInterval = 5.5
    private static let defaultRetryCount: Int = 3
    private static let defaultRetryInterval: TimeInterval = 0
    // Mirrors BNCPreferenceHelper's DEFAULT_THIRD_PARTY_APIS_TIMEOUT (500ms).
    private static let defaultThirdPartyAPIsWaitTime: TimeInterval = 0.5
    // Network timeout ceiling, matching the Android builder's 60s cap.
    private static let maxNetworkTimeout: TimeInterval = 60
    // Ceiling for the third-party API wait time, matching +[Branch setSDKWaitTimeForThirdPartyAPIs:].
    private static let maxThirdPartyAPIsWaitTime: TimeInterval = 10

    // MARK: - Identity & environment

    /// The Branch key (`key_live_...` or `key_test_...`). Required, must be non-empty.
    @objc public private(set) var branchKey: String

    /// When YES, Branch uses the test key found in your `Info.plist`. Default NO.
    @objc public var testMode: Bool = false

    /// Optional custom base URL for all calls to the Branch API. Must start with `http://` or `https://`.
    @objc public var apiUrl: String?

    /// Optional custom base URL for non-linking ("safe track") calls to the Branch API.
    @objc public var safeTrackAPIUrl: String?

    /// Optional custom base URL used for the CDN pattern list.
    @objc public var cdnBaseUrl: String?

    /// When YES, requests are routed to Branch's EU endpoints. Default NO.
    @objc public var euEndpoint: Bool = false

    // MARK: - Logging

    /// Log-level threshold for the Branch logging subsystem. Default `BranchLogLevelError`.
    @objc public var logLevel: BranchLogLevel = .error

    /// Optional custom log sink invoked for every emitted message at or above `logLevel`.
    @objc public var loggingCallback: BranchLogCallback?

    /// Optional per-request tracing hook (URL, request, response, error). Useful for debugging.
    @objc public var requestTracingCallback: callbackForTracingRequests?

    // MARK: - Network

    /// Request timeout in seconds. Must be > 0 and <= 60. Default matches the SDK default (5.5s).
    @objc public var networkTimeout: TimeInterval = BranchConfiguration.defaultNetworkTimeout

    /// Max number of retries on a Branch server error. Must be >= 0. Default 3.
    @objc public var retryCount: Int = BranchConfiguration.defaultRetryCount

    /// Seconds to wait between retries. Must be >= 0. Default 0.
    @objc public var retryInterval: TimeInterval = BranchConfiguration.defaultRetryInterval

    /// Optional network service class (must conform to `BNCNetworkServiceProtocol`).
    @objc public var remoteInterface: AnyClass?

    /// SDK wait time in seconds for third-party APIs. Must be > 0 and <= 10. Default 0.5s.
    @objc public var thirdPartyAPIsWaitTime: TimeInterval = BranchConfiguration.defaultThirdPartyAPIsWaitTime

    // MARK: - Privacy & attribution

    /// Optional consumer-protection attribution level. When nil, the SDK default is used.
    @objc public var attributionLevel: BranchAttributionLevel?

    /// When YES, limits Facebook attribution/tracking. Default NO.
    @objc public var limitFacebookAttribution: Bool = false

    /// When YES, disables callouts to ad networks for all events. Default NO.
    @objc public var adNetworkCalloutsDisabled: Bool = false

    /// Optional DMA (EEA) consent parameters. When nil, no DMA parameters are sent.
    @objc public var dmaParameters: BranchDMAParameters?

    // MARK: - URL collection

    private var mutableAllowedSchemes: [String] = []
    private var mutableUrlPatternsToIgnore: [String] = []
    private var mutableRequestMetadata: [String: String] = [:]

    /// URI schemes allowed to be tracked by Branch. Empty means all schemes (the default).
    @objc public var allowedSchemes: [String] { mutableAllowedSchemes }

    /// Regex patterns matching URLs that Branch should never transmit.
    ///
    /// Historically the Objective-C header declared this `readonly` while shipping an unreachable setter and
    /// documenting it as settable on `Branch.h`. It is now a genuinely settable property.
    @objc public var urlPatternsToIgnore: [String] {
        get { mutableUrlPatternsToIgnore }
        set { mutableUrlPatternsToIgnore = newValue }
    }

    // MARK: - Request metadata

    /// Key-value pairs included in the metadata on every request.
    @objc public var requestMetadata: [String: String] { mutableRequestMetadata }

    // MARK: - Open tracking

    /// When NO, Branch does not automatically send an open on foreground. Default YES.
    @objc public var automaticOpenEvents: Bool = true

    // MARK: - Pasteboard

    /// When YES, Branch checks the pasteboard for a Branch Link on app install. Default NO.
    @objc public var checkPasteboardOnInstall: Bool = false

    // MARK: - App Clip

    /// Optional App Group identifier used to share data between an App Clip and the full app.
    @objc public var appClipAppGroup: String?

    // MARK: - Debugging

    /// Optional constant parameters merged into every deep-link response, for debugging.
    @objc public var deepLinkDebugParams: [AnyHashable: Any]?

    /// The process-wide preference store. `BranchConfiguration` can seed itself from persisted values
    /// (`load(from:)`) and push its values into it (`apply(to:)`). Both are explicit — nothing is read
    /// or written implicitly at init/validate time, so a freshly created configuration keeps its
    /// documented defaults.
    private var preferenceHelper: BNCPreferenceHelper { BNCPreferenceHelper.sharedInstance() }

    // MARK: - Initialization

    /// Designated initializer.
    /// - Parameter branchKey: The Branch key. Must be non-empty; validated in `validate()`.
    @objc public init(key branchKey: String) {
        self.branchKey = branchKey
        super.init()
    }

    // MARK: - Factory helpers

    /// Debug preset: verbose logging and test mode enabled.
    @objc public class func debug(_ branchKey: String) -> BranchConfiguration {
        let config = BranchConfiguration(key: branchKey)
        config.logLevel = .verbose
        config.testMode = true
        return config
    }

    /// Production preset: warning-level logging.
    @objc public class func production(_ branchKey: String) -> BranchConfiguration {
        let config = BranchConfiguration(key: branchKey)
        config.logLevel = .warning
        return config
    }

    /// Compliance preset: minimal attribution level.
    @objc public class func compliance(_ branchKey: String) -> BranchConfiguration {
        let config = BranchConfiguration(key: branchKey)
        config.attributionLevel = .minimal
        return config
    }

    // MARK: - Mutators for collection / grouped settings

    /// Add a single URI scheme to the allow list.
    @objc public func addAllowedScheme(_ scheme: String) {
        mutableAllowedSchemes.append(scheme)
    }

    /// Add a single key/value pair to the per-request metadata.
    @objc public func addMetadata(key: String, value: String) {
        mutableRequestMetadata[key] = value
    }

    // MARK: - Validation

    /// Validates the configuration. Raises `NSInvalidArgumentException` with an actionable message if any
    /// field is invalid. Called automatically by `+[Branch initialize:]`.
    @objc public func validate() {
        if branchKey.isEmpty {
            raiseInvalidArgument("Branch key cannot be empty. Get your key from dashboard.branch.io/settings.")
        }
        if networkTimeout <= 0 {
            raiseInvalidArgument(String(format: "Network timeout must be a positive number of seconds (got %.2f).", networkTimeout))
        }
        if networkTimeout > BranchConfiguration.maxNetworkTimeout {
            raiseInvalidArgument(String(format: "Network timeout cannot exceed 60 seconds (got %.2f).", networkTimeout))
        }
        if retryCount < 0 {
            raiseInvalidArgument("Retry count must be >= 0 (got \(retryCount)).")
        }
        if retryInterval < 0 {
            raiseInvalidArgument(String(format: "Retry interval must be >= 0 seconds (got %.2f).", retryInterval))
        }
        if thirdPartyAPIsWaitTime <= 0 || thirdPartyAPIsWaitTime > BranchConfiguration.maxThirdPartyAPIsWaitTime {
            raiseInvalidArgument(String(format: "Third-party APIs wait time must be > 0 and <= 10 seconds (got %.2f).", thirdPartyAPIsWaitTime))
        }
        if let remoteInterface = remoteInterface,
           !remoteInterface.conforms(to: BNCNetworkServiceProtocol.self) {
            raiseInvalidArgument("remoteInterface class '\(NSStringFromClass(remoteInterface))' must conform to BNCNetworkServiceProtocol.")
        }
    }

    private func raiseInvalidArgument(_ reason: String) -> Never {
        NSException(name: .invalidArgumentException, reason: reason, userInfo: nil).raise()
        // NSException.raise() does not return; this is unreachable but satisfies the `Never` contract.
        fatalError(reason)
    }

    // MARK: - Preference-helper sync

    /// Writes the pre-init values that live in the shared preference store into it. Only the settings
    /// that `BNCPreferenceHelper` actually persists are pushed; URL/network endpoint and callback
    /// settings are applied elsewhere by `+[Branch initialize:]`. Call this explicitly — it is not
    /// invoked implicitly, so constructing a configuration never mutates process-wide state.
    @objc public func apply(to preferences: BNCPreferenceHelper) {
        preferences.timeout = networkTimeout
        preferences.retryCount = retryCount
        preferences.retryInterval = retryInterval
        preferences.thirdPartyAPIsWaitTime = thirdPartyAPIsWaitTime
        preferences.limitFacebookTracking = limitFacebookAttribution
        preferences.disableAdNetworkCallouts = adNetworkCalloutsDisabled

        if let attributionLevel = attributionLevel {
            // preferences.attributionLevel is declared as NSString; bridge explicitly.
            preferences.attributionLevel = attributionLevel as NSString
        }
        if let dmaParameters = dmaParameters {
            preferences.eeaRegion = dmaParameters.eeaRegion
            preferences.adPersonalizationConsent = dmaParameters.adPersonalizationConsent
            preferences.adUserDataUsageConsent = dmaParameters.adUserDataUsageConsent
        }
    }

    /// Convenience overload that targets the shared preference store.
    @objc public func applyToSharedPreferences() {
        apply(to: preferenceHelper)
    }

    /// Seeds this configuration from values already persisted in the preference store, so a caller can
    /// start from the SDK's current state and override selectively. Only fields the store round-trips
    /// are read; the rest keep their defaults.
    @objc public func load(from preferences: BNCPreferenceHelper) {
        networkTimeout = preferences.timeout
        retryCount = preferences.retryCount
        retryInterval = preferences.retryInterval
        thirdPartyAPIsWaitTime = preferences.thirdPartyAPIsWaitTime
        limitFacebookAttribution = preferences.limitFacebookTracking
        adNetworkCalloutsDisabled = preferences.disableAdNetworkCallouts

        // preferences.attributionLevel is an NSString; BranchAttributionLevel is an NS_STRING_ENUM
        // (a distinct Swift type), so bridge NSString -> String -> BranchAttributionLevel.
        let persistedLevel = preferences.attributionLevel as String?
        if let level = persistedLevel, !level.isEmpty {
            attributionLevel = BranchAttributionLevel(rawValue: level)
        }
        if preferences.eeaRegionInitialized() {
            dmaParameters = BranchDMAParameters.eeaRegion(preferences.eeaRegion,
                                                          adPersonalizationConsent: preferences.adPersonalizationConsent,
                                                          adUserDataUsageConsent: preferences.adUserDataUsageConsent)
        }
    }

    /// Convenience overload that reads from the shared preference store.
    @objc public func loadFromSharedPreferences() {
        load(from: preferenceHelper)
    }
}
